src/jdk.compiler/share/classes/com/sun/tools/javac/tree/DCTree.java

```diff
@@ -53,41 +84,150 @@
-     * This is primarily useful for nodes that wrap a single string child.
+     * {@return the start position of this tree node}
+     *
+     * For most nodes, this is the position of the first character that is unique to this node.
      *
-     * @param dc     the enclosing doc comment
-     * @param offset the offset
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
      */
-    public long getSourcePosition(DCDocComment dc, int offset) {
-        return dc.comment.getSourcePos(pos + offset);
+    public int getStartPosition() {
+        return pos;
     }
 
-    public JCDiagnostic.DiagnosticPosition pos(DCDocComment dc) {
-        return new SimpleDiagnosticPosition(dc.comment.getSourcePos(pos));
+    /**
+     * {@return the "preferred" position of this tree node}
+     *
+     * It is typically the position of the first character that is unique to this node.
+     * It is the position that is used for the caret in "line and caret" diagnostic messages.
+     *
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
+     */
+    public int getPreferredPosition() {
+        return pos;
     }
 
-    /** Convert a tree to a pretty-printed string. */
+    /**
+     * {@return the end position of the tree node}
+     *
+     * The value is typically derived in one of three ways:
+     * <ul>
+     * <li>computed from the start and length of "leaf" nodes, such as {@link TextTree},
+     * <li>computed recursively from the end of the last child node, such as for most {@link DCBlockTag block tags}, or
+     * <li>provided explicitly, such as for subtypes of {@link DCEndPosTree}
+     * </ul>
+     *
+     * The value is relative to the beginning of the comment text.
+     * Use {@link DCDocComment#getSourcePosition(int)} to convert
+     * it to a position in the source file.
+     */
+    public int getEndPosition() {
+        if (this instanceof DCEndPosTree<?> dcEndPosTree) {
+            int endPos = dcEndPosTree.getEndPos();
+
+            if (endPos != NOPOS) {
+                return endPos;
+            }
+        }
+
+        switch (getKind()) {
+            case TEXT -> {
+                DCText text = (DCText) this;
+                return text.pos + text.text.length();
+            }
+
+            case ERRONEOUS -> {
+                DCErroneous err = (DCErroneous) this;
+                return err.pos + err.body.length();
+            }
+
+            case IDENTIFIER -> {
+                DCIdentifier ident = (DCIdentifier) this;
+                return ident.pos + ident.name.length();
+            }
+
+            case AUTHOR, DEPRECATED, HIDDEN, PARAM, RETURN, SEE, SERIAL, SERIAL_DATA, SERIAL_FIELD, SINCE, THROWS, UNKNOWN_BLOCK_TAG, VERSION -> {
```

Comment:
You added HIDDEN and merged PARAM into this case of block tags. Should we also add PROVIDES and USES? If so, then we should add respective tests.

Reply:
Yes, and agreed.

Reply:
FWIW, I'm still in two minds whether this should be a switch or overriding methods. It's ... disappointing ... that the lack of support for `PROVIDES` and `USES` didn't "automatically" show up in testing.  That may suggest a lack of coverage in the `DocTree` tests, which suggests the need for a test to check coverage.

Reply:
For the sake of this review, using `switch` would be better: fewer changes, less cognitive load. Later we can think of ways of improving that code. It would be nice to structure it such that it would fail at compile time, if we forgot a case.

Reply:
It's hard to know how to make it fail at compile time.  There are essentially 3 groups of trees ... inline tags (and other trees that use `DCEndPosTree`), block tags (the big block of cases round about line 177), and "other".  We could arguably move block tags into the `default` case and use a type test, which would improve the future-safety.

As for tests, we *do* have tests for `ProvidesTree` and `UsesTree`, but the new `RangeChecker` didn't catch the missing case labels because the basic checks in `RangeChecker` were satisfied by the code in the default case in the switch statement.

Reply:
I agree that it might prove hard, if at all possible. That said, we should investigate this later, not in this PR. One potential direction for investigation would be "switching on sealed interfaces". A class or an interface can be a part of multiple sealed-interface hierarchies:

```
    sealed interface Tag permits Block, Inline {}
        sealed interface Block extends Tag permits Block1, Bimodal {}
            non-sealed interface Block1 extends Block {}
        sealed interface Inline extends Tag permits Inline1, Bimodal {}
            non-sealed interface Inline1 extends Inline {}
        non-sealed interface Bimodal extends Block, Inline {}
```

After pattern matching for `switch` (currently in preview via JEP 406) has been standardized, the exhaustiveness of a switch could help us catch bugs at compile time. Possibly.

Reply:
Maybe.